### PR TITLE
feat(conditional-layers): Implement generalized version of tri-layer support for "adjust" layer

### LINKED ISSUE #828: Feature request: Conditional layer support (including tri-layer for lower/raise/adjust)
I was thinking of simple ZMK features I could hack on that I'd actually use, and tri-layer support (like QMK and BlueMicro firmware offer) came to the top of my list. You can fake it today by nesting momentary layer mappings, but the layer deactivation behavior on layer up isn't the same. (For example, if you have press Lower, press Raise, release Lower, you stay on the "adjust" layer until _Raise_ is released. Real tri-layer support would take you back to the "raise" layer at this point.)

I thought about implementing this feature using new layer behaviors, or about integrating with the combo system somehow, but the cleanest approach I could come up with actually involves a new node next to the keymap like so:

```
#define LOWER 1
#define RAISE 2
#define ADJUST 3

multi_layers {
  compatible = "zmk,multi-layers";
  adjust {
    if-layers = <LOWER RAISE>;
    then-layer = ADJUST;
  };
};

keymap {
  compatible = "zmk,keymap";
  default_layer {
    &kp A &kp B
    &mo LOWER &mo RAISE
  } 
  lower_layer {
    &kp C &kp D
    &trans &trans
  }
  raise_layer {
    &kp E &kp F
    &trans &trans
  }
  adjust_layer {
    &kp G &kp H
    &trans &trans
  }
};
```

I wrote a [first draft implementation](https://github.com/zmkfirmware/zmk/pull/830) (that I've not tested much yet), and I'd like to get some early feedback in two areas. (Of course, it's also possible this approach is wrong on a more fundamental level. That feedback is welcome as well. 😄)

First, should multi-layer configs be applied recursively? In other words, should activating the then-layer of one config be able to trigger the if-layer of another config?

Advantages to recursive activation:

* Simpler implementation (this is how the event model works by default)
* More flexible for users

Disadvantages to recursive activation:

* Possible for a circular reference between multi-layer configs to cause permanent layer activation, since two layers could reactivate one another immediately after deactivating.

Second, should multi-layer (de)activation update the existing layer state bitset, or should there be a new bitset for multi-layer activations that gets OR'd with the existing layer state bitset to compute the "effective layers state"?

Advantages to updating the existing layers state:

* Simpler implementation (no changes needed to the core layers impl for this feature)

Disadvantages to updating the existing layers state:

* Interactions with existing layer activation mechanisms may be confusing, e.g., if a layer is simultaneously activated via both a `&mo` behavior and a multi-layer config, ideally it would be deactivated only after _both_ activation conditions stop triggering, but if the two approaches share a layer bitset, the layer will become inactive earlier, when _the first_ activation condition stops.

### PR DESCRIPTION
**Pull request by bcat**
I missed tri-layer support from QMK, so I implemented a ZMK version. It seemed most logical to express as an N-to-1 mapping of `if-layers` to activate a single `then-layer`, so I called it "conditional layers".

As discussed in the attached issue, I considered calling this "layer combos" and replacing the `then-layer` property with a `bindings` property that could trigger an `&mo` behavior (or any other behavior as well), but in practice I didn't see any reason to generalize that much.

Fixes #828.

### Tasks before review

- [X] Initial implementation
- [X] Basic unit tests
- [x] Clarity on two TODOs in the code (see linked issue for more details)
- [x] Documentation
- [x] Unit tests for different orderings of layer down/up
- [x] Unit tests for multiple conditional layer configs in the same keymap
- [x] Testing on real keyboard

### CODE REVIEW

**Comment by caksoylar** [comment]
Hey @bcat, I was missing tristate layers from my QMK config thinking about implementing myself until I found this. This PR works pretty well for me at the moment and it would be great to have this available in ZMK. You bring up some good points for discussion in https://github.com/zmkfirmware/zmk/issues/828 and the TODO items in the code, but IMO it would still be a very good addition with all the current caveats (i.e. do not define chains of multi-layers, do not activate the then layer through multiple methods simultaneously); I am not sure whether/how QMK handles those issues either.

PS: I kind of like the `conditional layers` name better than `multi layers` as well; it fits the if-then terminology too.

**Comment by bcat** [comment]
Thanks for testing it! I sort of let this stagnate both because I wasn't sure if anyone else had a use for it and because life has just been busy lately. But I may have some time this weekend to update the PR (e.g., finish up the unit tests) and put it out for a real review.

I do agree on "conditional layers" as a clearer name. That should be easy enough to change. And I agree the edge cases are probably fine for now. The big open question for me is the second point on https://github.com/zmkfirmware/zmk/issues/828#issuecomment-860344899. Namely should the "then layer" just use a normal binding specification (like combos do). That seems a bit cleaner to me. If I have some cycles this weekend I might give that a try.

**Comment by caksoylar** [comment]
> Namely should the "then layer" just use a normal binding specification (like combos do). 

I don't have much of a preference either way, but I can't think of a useful behavior other than `&mo` like you said in your comment. If that is the case, I would tend towards keeping it simple as it is.

**Comment by bcat** [comment]
> > Namely should the "then layer" just use a normal binding specification (like combos do).
> 
> I don't have much of a preference either way, but I can't think of a useful behavior other than `&mo` like you said in your comment. If that is the case, I would tend towards keeping it simple as it is.

Yeah, I think `&mo` is the only practically useful binding there. IIRC, I mostly liked the idea of using bindings because it sidestepped the problem of imbalanced layer activate/deactive actions (since the actual manipulation of the layer state was done by the standard, existing behavior). But honestly I think it depends if it makes the implementation more complex or not. I'll have to take a closer look at the code and see.

**Comment by bcat** [comment]
Rebased against latest main, renamed to "conditional layers", and completed docs/tests. Tomorrow I'll make sure this still works on a physical keyboard, and then I'll send it out for review.

**Comment by bcat** [comment]
Seems to still work fine on my Crkbd (with nice!nano v1). Should be ready for review now. :)

@caksoylar, mind giving the latest version of this PR a try when you have a chance and verifying I haven't broken anything for you? Thanks!

**Comment by bcat** [comment]
Here's an example keymap that uses this feature: https://github.com/bcat/zmk-config/commit/63f4dbb61c9d5fd8ff94273b09965d63b8fd6590

**Comment by caksoylar** [comment]
I updated my fork with the PR, then [readjusted my keymap](https://github.com/caksoylar/zmk-config/commit/e19e1df4b914441151c961a6be136506655880c9) for the latest changes. Still working great here, thanks for your work!

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:42]: I had some trouble parsing this statement. Does this mean that an infinite loop isn't technically possible? I assume so, if `if-layers` is constrained to have at least two elements.

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:5]:
```suggestion
Conditional layers support activates a particular layer (called the `then-layer`) when all layers in a
```

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:6]:
```suggestion
specified set (called the `if-layers`) are active. This feature generalizes what's commonly known as
```

File: docs/docs/features/conditional-layers.md
Comment (bcat) [reply docs/docs/features/conditional-layers.md]: Yeah, as far as a I know, an infinite loop shouldn't be possible since the implementation only tries to activate a layer if it's not already active, and so in the worst case you end up done when every layer is active. But honestly, on further thought I don't think that detail is really important, so I pruned down the note text. WDYT?

Comment (caksoylar) [reply docs/docs/features/conditional-layers.md]: I think that's good, the user doesn't really need to know whether it can cause an issue anyway, especially if it cannot :)

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:5]:
```suggestion
Conditional layers support activating a particular layer (called the `then-layer`) when all layers in
```

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:29]:
```suggestion
Each conditional layer configuration may have whatever name you like, but it's helpful to choose
```

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:32]:
```suggestion
- `if-layers` specifies a set of layer numbers, all of which must be active for the
```

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:34]:
```suggestion
- `then-layer` specifies a layer number that should be activated if and only if all the
```

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:37]:
```suggestion
Therefore, in this example, layer 3 ("adjust") will be activated if and only if both layers 1
```

File: docs/docs/features/conditional-layers.md
Comment (caksoylar) [inline docs/docs/features/conditional-layers.md:51]: Can we add a note similar to the point 2. in [QMK docs](https://docs.qmk.fm/#/ref_functions?id=update_tri_layer_statestate-x-y-z), saying that the `then-layer` should be higher than the `if-layers` to be able to access it reliably?

**Review by caksoylar — COMMENTED** [review]
I noted a few language improvements that I missed the first time around, and suggestion to add a remark regarding layer order in the docs.

**Comment by bcat** [comment]
Thanks! I've incorporated those suggestions as well.

BTW, I'm not used to reviews cycles in PRs. Instead of accumulating lots of tiny commits, should I squash them into more logical chunks and force push? I'm happy to do whatever is preferred. (I know QMK maintainers often squash the whole PR into one commit when it's merged, so it's a moot point, but I wasn't sure if ZMK did the same.)

**Review by caksoylar — APPROVED** [review]
Thanks for incorporating the suggestions! I can approve in terms of docs, but other contributors will need to review. 

Regarding commits, AFAIK PRs can be merged as is as long as individual commits make sense, but they can get squashed otherwise. If you feel it can be improved you can reorganize/rebase them but I wouldn't worry about it too much.

**Comment by caksoylar** [comment]
I also haven't had any issues since I started using it last week.

**Review by petejohanson — APPROVED** [review]
This is perfect! Can you rebase to fix the merge conflict so we can merge?

Thanks!

**Comment by bcat** [comment]
> This is perfect! Can you rebase to fix the merge conflict so we can merge?

Thanks for the review! Rebased (and squashed some of doc revision commits while I was in there) and pushed. All the tests still pass, though as mentioned on Discord, I don't currently have a presently have a working ZMK board available to test on real hardware.

**Comment by bcat** [comment]
> This is perfect! Can you rebase to fix the merge conflict so we can merge?

Thanks for the review! Rebased (and squashed some of doc revision commits while I was in there) and pushed. All the tests still pass, though as mentioned on Discord, I don't currently have a presently have a working ZMK board available to test on real hardware.

**Comment by caksoylar** [comment]
I have been rebasing this on `main` and using it for the last the couple of months, I can confirm that the newly pushed version works fine as well.

**Comment by bcat** [comment]
Of course I managed to screw up a one-line merge. :) Fixed `sidebar.js` now, for realz.

**Review by petejohanson — APPROVED** [review]

**Comment by jordiorlando** [comment]
I'm thrilled that this has been merged, but I'm seeing an issue in a corner case. [My keymap](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) has layer-tap space keys for `raise` and `lower`, and I can't get the `adjust` layer to activate with this configuration. Any advice?

**Comment by bcat** [comment]
> I'm thrilled that this has been merged, but I'm seeing an issue in a corner case. [My keymap](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) has layer-tap space keys for `RAISE` and `LOWER`, and I can't get the `CONFIG` layer to activate with this configuration. Any advice?

Oh, that is interesting. The intent of the new functionality was that it would work with any type of layer binding (since the implementation just listens for layer state changes rather than monitoring individual behaviors). But I can definitely believe that I missed some edge case. :)

The next few days are likely to be busy for me, unfortunately, but I can try to isolate a unit test that repros the LT bug (if nobody else beats me to it).

**Comment by bcat** [comment]
So I did the simplest, quickest thing and changed https://github.com/zmkfirmware/zmk/blob/30ed2375335acf47da0216c6832ecfd30ea7097c/app/tests/conditional-layer/tri-layer/native_posix.keymap locally to use `&lt` instead of `&mo` (lengthening the `ZMK_MOCK_PRESS` durations to ensure the "hold" triggers). But that modified test still passes. So it's not as simple as "layer taps don't work with conditional layers". @jordiorlando, I think there's some more complex interaction with different parts of your keymap going on here. I'm not sure exactly what yet, though. Will need to dig a bit...

**Comment by jcmkk3** [comment]
> I'm thrilled that this has been merged, but I'm seeing an issue in a corner case. [My keymap](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) has layer-tap space keys for `RAISE` and `LOWER`, and I can't get the `CONFIG` layer to activate with this configuration. Any advice?

Have you tried to switch your LT to be MO just to ensure that the conditional-layer functionality is working for you?

**Comment by jordiorlando** [comment]
> > I'm thrilled that this has been merged, but I'm seeing an issue in a corner case. [My keymap](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) has layer-tap space keys for `RAISE` and `LOWER`, and I can't get the `CONFIG` layer to activate with this configuration. Any advice?
> 
> Have you tried to switch your LT to be MO just to ensure that the conditional-layer functionality is working for you?

Good idea, it's possible my build isn't pulling the latest version of ZMK. Trying that now.

**Comment by bcat** [comment]
Indeed, I think that is the issue. The `west.yml` file for this config points at a third-party fork of ZMK: https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/a521a785453668adba0678df668b81436cd29532/config/west.yml. Specifically, this fork: https://github.com/LOWPROKB/zmk. Which seems a bit outdated (last commit there is e6645882c32454251466949fa9a6447b195acd72, before the conditional layers PR was merged).

Is this keyboard supported in upstream ZMK? If so, you can probably just switch your config repo to upstream ZMK. But if not, sadly I think you'll need to create your own fork of the third-party repo and manually integrate this PR for the time being.

@petejohanson, it looks like the "Corne-ish Zen" maker's sample repo points at a nonstandard ZMK fork (https://github.com/LOWPROKB/zmk-config-Corne-ish-Zen/blob/b8de4f2b6545712141cc4f4c19260eaae89d2245/config/west.yml), so I suspect lots of folks are copying that by default. :( Is that a supported practice? I honestly don't mean to stir up drama (and will happily shut up if this is an existing issue), but it seems like this may lead to lots of end users building off ZMK forks with various degrees of bitrot....

**Comment by bcat** [comment]
Somewhat separately... I'm surprised the build doesn't just fail if there's unknown stuff in the keymap file. I don't really know how devicetree works under the hood, but is there some way to guard against this at build time rather than just silently ignoring the unsupported bits of the keymap?

**Comment by jordiorlando** [comment]
This would be a good question for @LOWPROKB as well. I think there was talk of upstreaming support for the Corne-ish Zen, but I don't know the current status.

**Comment by bcat** [comment]
Indeed, that's fair. Also, as it looks like there is not a bug with the conditional layers feature itself, maybe we should move the discussion about upstreaming, forks, etc. to a new issue? (Not trying to be too pedantic, it just seems like the root cause is not related.)

**Comment by jordiorlando** [comment]
> Indeed, that's fair. Also, as it looks like there is not a bug with the conditional layers feature itself, maybe we should move the discussion about upstreaming, forks, etc. to a new issue? (Not trying to be too pedantic, it just seems like the root cause is not related.) But then again, I don't really know _where_ a better to place to discuss would be....

Agreed, I don't want to flood this thread with off-topic jabber. Thank you for the super quick help troubleshooting!

I created a new issue over [here](https://github.com/LOWPROKB/zmk-config-Corne-ish-Zen/issues/3) that is probably more suitable for discussion.

**Comment by jordiorlando** [comment]
Just wanted to leave a final comment to say that I've tested [my layout](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) with the latest version of zmk, and everything works as intended. So stoked to have this feature! Thanks all.

**Comment by petejohanson** [comment]
> Indeed, I think that is the issue. The `west.yml` file for this config points at a third-party fork of ZMK: https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/a521a785453668adba0678df668b81436cd29532/config/west.yml. Specifically, this fork: https://github.com/LOWPROKB/zmk. Which seems a bit outdated (last commit there is [e664588](https://github.com/zmkfirmware/zmk/commit/e6645882c32454251466949fa9a6447b195acd72), before the conditional layers PR was merged).
> 
> Is this keyboard supported in upstream ZMK? If so, you can probably just switch your config repo to upstream ZMK. But if not, sadly I think you'll need to create your own fork of the third-party repo and manually integrate this PR for the time being.
> 
> @petejohanson, it looks like the "Corne-ish Zen" maker's sample repo currently points at a nonstandard ZMK fork (https://github.com/LOWPROKB/zmk-config-Corne-ish-Zen/blob/b8de4f2b6545712141cc4f4c19260eaae89d2245/config/west.yml), so I suspect lots of folks are copying that by default. :( Is that a supported practice? I honestly don't mean to stir up drama (and will happily shut up if this is already a known issue), but it seems like this may lead to lots of end users building off ZMK forks with various degrees of bitrot....

This is an unfortunate situation, certainly, and something that the modular work should help formalize. The battery changes in particular are the main holdup.

**Comment by bcat** [comment]
> > Indeed, I think that is the issue. The `west.yml` file for this config points at a third-party fork of ZMK: https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/a521a785453668adba0678df668b81436cd29532/config/west.yml. Specifically, this fork: https://github.com/LOWPROKB/zmk. Which seems a bit outdated (last commit there is [e664588](https://github.com/zmkfirmware/zmk/commit/e6645882c32454251466949fa9a6447b195acd72), before the conditional layers PR was merged).
> > Is this keyboard supported in upstream ZMK? If so, you can probably just switch your config repo to upstream ZMK. But if not, sadly I think you'll need to create your own fork of the third-party repo and manually integrate this PR for the time being.
> > @petejohanson, it looks like the "Corne-ish Zen" maker's sample repo currently points at a nonstandard ZMK fork (https://github.com/LOWPROKB/zmk-config-Corne-ish-Zen/blob/b8de4f2b6545712141cc4f4c19260eaae89d2245/config/west.yml), so I suspect lots of folks are copying that by default. :( Is that a supported practice? I honestly don't mean to stir up drama (and will happily shut up if this is already a known issue), but it seems like this may lead to lots of end users building off ZMK forks with various degrees of bitrot....
> 
> This is an unfortunate situation, certainly, and something that the modular work should help formalize. The battery changes in particular are the main holdup.

Makes sense. I guess the only thing I'd like to find out then is whether it's expected that there was no build failure here. Maybe that's just how devicetree works, but it kind of surprises me.

**Comment by petejohanson** [comment]
> 
> Makes sense. I guess the only thing I'd like to find out then is whether it's expected that there was no build failure here. Maybe that's just how devicetree works, but it kind of surprises me.

Yeah, unfortunately, the current build doesn't prevent this. I do have some ideas for how to improve this.

**Pull request merged**
